src/util/sandbox/fd_sandbox.c

```diff
@@ -0,0 +1,256 @@
+#ifdef FD_HAS_SANDBOX
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "fd_sandbox.h"
+#include "fd_sandbox_private.h"
+
+#include <errno.h>        /* errno */
+#include <fcntl.h>        /* open */
+#include <linux/audit.h>
+#include <linux/filter.h>
+#include <linux/seccomp.h>
+#include <sched.h>        /* CLONE_*, setns, unshare */
+#include <stddef.h>
+#include <stdio.h>        /* snprintf */
+#include <stdlib.h>       /* clearenv, mkdtemp*/
+#include <sys/mount.h>    /* MS_*, MNT_*, mount, umount2 */
+#include <sys/prctl.h>
+#include <sys/resource.h> /* RLIMIT_*, rlimit, setrlimit */
+#include <sys/stat.h>     /* mkdir */
+#include <sys/syscall.h>  /* SYS_* */
+#include <unistd.h>       /* set*id, sysconf, close, chdir, rmdir syscall */
+
+#include "../log/fd_log.h"
+
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
+
+void
+fd_sandbox_profile_init( fd_sandbox_profile_t * profile ) {
+  profile->initialized      = true;
+  profile->chroot_path      = NULL;
+  profile->netns            = "fd-frank-netless",
+  profile->seccomp_prog     = NULL;
+  profile->user             = "nobody";
+  profile->close_fds_beyond = 3U;
+  profile->max_open_fds     = 3U;
+}
+
+int
+fd_sandbox( const fd_sandbox_profile_t * profile ) {
+  if ( FD_UNLIKELY( !profile->initialized ) ) {
+    FD_LOG_ERR(( "the sandbox profile has not been initialized" ));
+  }
+
+  // Get target user before doing other kinds of restrictions
+  struct passwd * uinfo = getpwnam( profile->user );
+  if ( FD_UNLIKELY( !uinfo ) ) {
+    FD_LOG_ERR(( "getpwnam: %s", strerror( errno ) ));
+  }
+
+  // todo: maybe not clearenv but only keep vars prefixed with FD_
+  clearenv();
+  fd_sandbox_setup_netns( profile->netns );
+  fd_sandbox_setup_mountns( profile->chroot_path );
+  fd_sandbox_set_resource_limits( profile->max_open_fds );
+  fd_sandbox_change_user( uinfo );
+  fd_sandbox_close_fds_beyond( profile->close_fds_beyond );
+  fd_sandbox_seccomp( profile->seccomp_prog );
+
+  FD_LOG_INFO(( "thread group sandboxed" ));
+  return 0;
+}
+
+void
+fd_sandbox_change_user( struct passwd * u ) {
+  errno = 0;
+  
+  if ( FD_UNLIKELY( u->pw_uid == 0 ) ) {
+    FD_LOG_ERR(( "will not run as root" ));
+  }
+
+  FD_LOG_INFO(( "switching to user='%s' uid='%d' gid='%d'", u->pw_name, u->pw_uid, u->pw_gid ));
+
+  // setregid before setreuid otherwise setregid won't be allowed
+
+  if ( FD_UNLIKELY( setregid( u->pw_gid, u->pw_gid ) ) ) {
+    FD_LOG_ERR(( "setregid: %s", strerror( errno ) ));
+  }
+
+  if ( FD_UNLIKELY( setreuid( u->pw_uid, u->pw_uid ) ) ) {
+    FD_LOG_ERR(( "setreuid: %s", strerror( errno ) ));
+  }
+}
```

Comment:
See notes about fd_log_user.  No objection here but how fd_log_user behaves in conjunction with this API should be sorted.